### PR TITLE
[WIP] Add zalgo-promise

### PR DESCRIPTION
**Pull request by daniel-shuy**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

Select one of these and delete the others:

If adding a new definition:
- [x] The package does not already provide its own types, or cannot have its `.d.ts` files generated via `--declaration`
- [x] If this is for an npm package, match the name. If not, do not conflict with the name of an npm package.
- [x] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [x] Represents shape of module/library [correctly](https://www.typescriptlang.org/docs/handbook/declaration-files/library-structures.html)
- [x] `tslint.json` [should contain](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#linter-tslintjson) `{ "extends": "dtslint/dt.json" }`, and no additional rules.
- [x] `tsconfig.json` [should have](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#tsconfigjson) `noImplicitAny`, `noImplicitThis`, `strictNullChecks`, and `strictFunctionTypes` set to `true`.

If changing an existing definition:
- [ ] Provide a URL to documentation or source code which provides context for the suggested changes: <<url here>>
- [ ] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.

If removing a declaration:
- [ ] If a package was never on Definitely Typed, you don't need to do anything. (If you wrote a package and provided types, you don't need to register it with us.)
- [ ] Delete the package's directory.
- [ ] Add it to `notNeededPackages.json`.

### CODE REVIEW

File: types/zalgo-promise/promise.d.ts [hunk 1]
```diff
@@ -0,0 +1,56 @@
+type ResultType<P extends ZalgoPromise<any>> = P extends ZalgoPromise<infer R> ? R : never;
+
+type FlattenPromises<T extends {}> = {
+    [K in keyof T]: T[K] extends ZalgoPromise<any> ? ResultType<T[K]> : T[K];
+};
```
Comment (daniel-shuy) [inline types/zalgo-promise/promise.d.ts:5]: Should I export these 2 types, inline them, or disable automatic exporting?

File: types/zalgo-promise/zalgo-promise-tests.ts [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+const promisesObject = {
+    booleanPromise: ZalgoPromise.resolve(false),
+    boolean: false,
+    numberPromise: ZalgoPromise.resolve(0),
+    number: 0,
+    stringPromise: ZalgoPromise.resolve(''),
+    string: '',
+};
+// $ExpectType ZalgoPromise<{ booleanPromise: boolean, boolean: boolean, numberPromise: number, number: number, stringPromise: string, string: string }>
+ZalgoPromise.hash(promisesObject);
```
Comment (daniel-shuy) [inline types/zalgo-promise/zalgo-promise-tests.ts:79]: Anyone knows why returning `ZalgoPromise<FlattenPromises<...>>` in `ZalgoPromise.all()` correctly expands the tuple to `ZalgoPromise<[boolean, boolean, number, number, string, string]>` , but `ZalgoPromise.hash()` fails to expand the object to `ZalgoPromise<{ booleanPromise: boolean, boolean: boolean, numberPromise: number, number: number, stringPromise: string, string: string }>`?

```
expect                           TypeScript@4.5 expected type to be:
  ZalgoPromise<{ booleanPromise: boolean, boolean: boolean, numberPromise: number, number: number, stringPromise: string, string: string }>
got:
  ZalgoPromise<FlattenPromises<{ booleanPromise: ZalgoPromise<boolean>; boolean: boolean; numberPromise: ZalgoPromise<number>; number: number; stringPromise: ZalgoPromise<string>; string: string; }>>
```

File: types/zalgo-promise/zalgo-promise-tests.ts [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+};
+// $ExpectType ZalgoPromise<{ booleanPromise: boolean, boolean: boolean, numberPromise: number, number: number, stringPromise: string, string: string }>
+const hash: ZalgoPromise<{
+    booleanPromise: boolean,
+    boolean: boolean,
+    numberPromise: number,
+    number: number,
+    stringPromise: string,
+    string: string
+}> = ZalgoPromise.hash(promisesObject);
```
Comment (daniel-shuy) [inline types/zalgo-promise/zalgo-promise-tests.ts:86]: How can I assert the variable (`hash`) type instead of the function call return type?

Comment (Nadhum) [reply types/zalgo-promise/zalgo-promise-tests.ts:86]: // $ExpectType type
hash;

Comment (daniel-shuy) [reply types/zalgo-promise/zalgo-promise-tests.ts:86]: @Nadhum thanks!

**Comment by daniel-shuy** [comment]
Since this PR has gotten quite messy, I'll close it and reopen a new PR

**Pull request closed without merging**